be/src/storage/lake/vacuum.cpp

```diff
@@ -631,4 +633,210 @@ void delete_txn_log(TabletManager* tablet_mgr, const DeleteTxnLogRequest& reques
+                                                                    return true;
+                                                                })),
+                              "Failed to list " + segment_root_location);
+    LOG(INFO) << "Listed all data files, total files: " << total_files << ", total bytes: " << total_bytes
+              << ", candidate files: " << data_files.size();
+    return data_files;
+}
+
+static StatusOr<std::map<std::string, DirEntry>> find_orphan_data_files(FileSystem* fs, std::string_view root_location,
+                                                                        int64_t expired_seconds,
+                                                                        std::ostream& audit_ostream) {
+    const auto metadata_root_location = join_path(root_location, kMetadataDirectoryName);
+    const auto segment_root_location = join_path(root_location, kSegmentDirectoryName);
+
+    ASSIGN_OR_RETURN(auto metadatas, list_tablet_metadata(fs, metadata_root_location));
+
+    if (metadatas.empty()) {
+        LOG(INFO) << "Skipped data file GC of " << root_location << ", because there is no tablet metadata";
+        return std::map<std::string, DirEntry>();
+    }
+
+    ASSIGN_OR_RETURN(auto data_files, list_data_files(fs, segment_root_location, expired_seconds));
+
+    if (data_files.empty()) {
+        return data_files;
+    }
+
+    std::set<std::string> data_files_in_metadatas;
+    auto check_rowset = [&](const RowsetMetadata& rowset) {
+        for (const auto& seg : rowset.segments()) {
+            data_files.erase(seg);
+            data_files_in_metadatas.emplace(seg);
+        }
+    };
+
+    audit_ostream << "Total metadatas: " << metadatas.size() << std::endl;
+    LOG(INFO) << "Start to filter with metadatas, count: " << metadatas.size();
+
+    int64_t progress = 0;
+    for (const auto& [name, metadata] : metadatas) {
+        for (const auto& rowset : metadata->rowsets()) {
+            check_rowset(rowset);
+        }
+        ++progress;
+        audit_ostream << '(' << progress << '/' << metadatas.size() << ") " << name << '\n'
+                      << proto_to_json(*metadata) << std::endl;
+        LOG(INFO) << "Filtered with meta file: " << name << " (" << progress << '/' << metadatas.size() << ')';
+    }
+
+    LOG(INFO) << "Start to double checking";
+
+    for (const auto& [name, entry] : data_files) {
+        if (data_files_in_metadatas.contains(name)) {
+            LOG(WARNING) << "Failed to do double checking, file: " << name;
+            return Status::InternalError("Failed to do double checking");
+        }
+    }
+
+    LOG(INFO) << "Succeed to do double checking";
+    LOG(INFO) << "Found " << data_files.size() << " orphan files";
+
+    return data_files;
+}
+
+// root_location is a partition dir in s3
+Status datafile_gc(std::string_view root_location, std::string_view audit_file_path, int64_t expired_seconds,
```

Comment:
what if the root_location is specified wrongly, e.g. to the xxx/<serviceid>/<table> will all the segment files been treated as garbage?

Reply:
root_location need data&meta subdir